src/fs.rs

```diff
@@ -418,40 +494,40 @@ impl Fs {
 
     /// Like [std::path::Path::is_file] but for the guest filesystem.
     pub fn is_file(&self, path: &GuestPath) -> bool {
-        matches!(self.lookup_node(path), Some(FsNode::File { .. }))
+        matches!(self.lookup_node(path), Some(FsNode::HostFile { .. }))
     }
 
     /// Like [std::fs::read] but for the guest filesystem.
     pub fn read<P: AsRef<GuestPath>>(&self, path: P) -> Result<Vec<u8>, ()> {
-        let node = self.lookup_node(path.as_ref()).ok_or(())?;
-        let FsNode::File {
-            host_path,
-            writeable: _,
-        } = node else {
-            return Err(())
-        };
-        Ok(handle_open_err(std::fs::read(host_path), host_path))
+        let mut file = self.open(path.as_ref())?;
+        let mut result = Vec::new();
+        file.read_to_end(&mut result).map_err(|_| ())?;
+        Ok(result)
     }
 
     /// Like [std::fs::File::open] but for the guest filesystem.
     #[allow(dead_code)]
-    pub fn open<P: AsRef<GuestPath>>(&self, path: P) -> Result<std::fs::File, ()> {
+    pub fn open<P: AsRef<GuestPath>>(&self, path: P) -> Result<File, ()> {
+        // it would be nice to delegate to self.open_with_options, but currently it wants a mutable reference to self
```

Comment:
Ah, so you ran into this too, that's indeed exactly why I have the two implementations. I'm a bit scared they'll diverge in behaviour in some corner case eventually, but not requiring `&mut` is nice. Maybe I should write some tests for this code, it's some of the trickiest stuff to get right and also the easiest to test.

(Just dismiss this, I'm not asking you to change the comment, it's good!)

Reply:
It might be a good idea to have interior mutability in fs, so it would always take a `&self`?

Reply:
I think I would like it if you need `&mut` to “mutate” the FS somehow, but interior mutability would be a nice trick to solve the implementation-sharing problem. We can enforce that invariant within a wrapper function in that case. Changing this should be a separate commit or PR though.

Reply:
Sure